R/functions__eager.R

```diff
@@ -37,41 +38,88 @@
 #' # diagonal
 #' pl$concat(l_hor, how = "diagonal")
 pl$concat = function(
```

Comment:
There should be some checks that inputs are of the same class. Currently, `pl$concat(pl$DataFrame(mtcars), pl$LazyFrame(mtcars))` works while it shouldn't. 

`py-polars` errors in this case:
```
pl.concat([pl.DataFrame({"x": 1}), pl.LazyFrame({"x": 1})])

AttributeError: 'LazyFrame' object has no attribute '_df'. Did you mean: '_ldf'?      
```

Reply:
I would argue that is a good thing, quite R like, and it does not break anything.

We use robj_to!() to decide if some Robj meaningfully can be coerced into another R/rust object, otherwise an appropriate error is returned/raised.

```R
# cbind a data.frame with a vector in R
> cbind(mtcars, zen_value = runif(32))
                     mpg cyl  disp  hp drat    wt  qsec vs am gear carb  zen_value
Mazda RX4           21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4 0.94224509
Mazda RX4 Wag       21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4 0.25997295
Datsun 710          22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1 0.70630125
Hornet 4 Drive      21.4   6 258.0 110 3.08 3.215 19.44  1  0    3    1 0.75788695
```


```R
# pl$concat DataFrame with Series horizontally
> pl$concat(pl$DataFrame(mtcars),pl$Series(runif(32),"zen-value"),how = "horizontal")
shape: (32, 12)
┌──────┬─────┬───────┬───────┬───┬─────┬──────┬──────┬───────────┐
│ mpg  ┆ cyl ┆ disp  ┆ hp    ┆ … ┆ am  ┆ gear ┆ carb ┆ zen-value │
│ ---  ┆ --- ┆ ---   ┆ ---   ┆   ┆ --- ┆ ---  ┆ ---  ┆ ---       │
│ f64  ┆ f64 ┆ f64   ┆ f64   ┆   ┆ f64 ┆ f64  ┆ f64  ┆ f64       │
╞══════╪═════╪═══════╪═══════╪═══╪═════╪══════╪══════╪═══════════╡
│ 21.0 ┆ 6.0 ┆ 160.0 ┆ 110.0 ┆ … ┆ 1.0 ┆ 4.0  ┆ 4.0  ┆ 0.045261  │
```


```R
> pl$concat(pl$LazyFrame(mtcars),mtcars)
[1] "polars LazyFrame naive plan: (run ldf$describe_optimized_plan() to see the optimized plan)"
UNION
  PLAN 0:
    DF ["mpg", "cyl", "disp", "hp"]; PROJECT */11 COLUMNS; SELECTION: "None"
  PLAN 1:
    DF ["mpg", "cyl", "disp", "hp"]; PROJECT */11 COLUMNS; SELECTION: "None"
END UNION
> pl$concat(mtcars, pl$LazyFrame(mtcars))
shape: (64, 11)
┌──────┬─────┬───────┬───────┬───┬─────┬─────┬──────┬──────┐
│ mpg  ┆ cyl ┆ disp  ┆ hp    ┆ … ┆ vs  ┆ am  ┆ gear ┆ carb │
│ ---  ┆ --- ┆ ---   ┆ ---   ┆   ┆ --- ┆ --- ┆ ---  ┆ ---  │
│ f64  ┆ f64 ┆ f64   ┆ f64   ┆   ┆ f64 ┆ f64 ┆ f64  ┆ f64  │
╞══════╪═════╪═══════╪═══════╪═══╪═════╪═════╪══════╪══════╡
│ 21.0 ┆ 6.0 ┆ 160.0 ┆ 110.0 ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0  ┆ 4.0  │
│ 21.0 ┆ 6.0 ┆ 160.0 ┆ 110.0 ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0  ┆ 4.0  │
│ 22.8 ┆ 4.0 ┆ 108.0 ┆ 93.0  ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0  ┆ 1.0  │
.....

> rbind(data.table(mtcars),mtcars) |> str()
Classes ‘data.table’ and 'data.frame':	64 obs. of  11 variables:
 $ mpg : num  21 21 22.8 21.4 18.7 18.1 14.3 24.4 22.8 19.2 ...
 $ cyl : num  6 6 4 6 8 6 8 4 4 6 ...
.
.
.
 $ am  : num  1 1 1 0 0 0 0 0 0 0 ...
 $ gear: num  4 4 4 3 3 3 3 4 4 4 ...
 $ carb: num  4 4 1 1 2 1 4 2 2 4 ...
 - attr(*, ".internal.selfref")=<externalptr> 
> rbind(mtcars,data.table(mtcars)) |> str()
'data.frame':	64 obs. of  11 variables:
 $ mpg : num  21 21 22.8 21.4 18.7 18.1 14.3 24.4 22.8 19.2 ...
 $ cyl : num  6 6 4 6 8 6 8 4 4 6 ...
 $ disp: num  160 160 108 258 360 ...

```

Reply:
I thought we should mimic `py-polars` as much as possible. This automatic collecting can lead to unexpected output. For example if I want to concat several LazyFrames at once, but I forgot the first one of them is a DataFrame, then all other LazyFrames are collected, which can take some time. I just think it's weird that this function doesn't only concatenate but collect as well, especially since it's the only one (I think?) that collects automatically.

I'm fine with appending Series to a DataFrame, but I'd rather error if DataFrames and LazyFrames are mixed in the input. Or maybe adding an argument `auto_collect` which would be `FALSE` by default?

Reply:
You're right, there should be no implicit auto collect. I actually disliked that from spark, I was not always sure if a collect would be triggered.